quark-test.c

```diff
@@ -0,0 +1,334 @@
+
+	return (child);
+}
+
+static int
+drain_for_pid(struct quark_queue *qq, pid_t pid, struct quark_event *qev)
+{
+	int			n;
+
+	for (;;) {
+		n = quark_queue_get_events(qq, qev, 1);
+		if (n == -1) {
+			err(1, "quark_queue_get_events");
+		} else if (n == 0) {
+			if (quark_queue_block(qq) == -1)
+				err(1, "quark_queue_block");
+			continue;
+		} else if (n != 1)
+			errx(1, "quark_queue_get_events is broken");
+
+		if (qev->process == NULL)
+			continue;
+		if (qev->process->pid != (u32)pid)
+			continue;
+		break;
+	}
+
+	return (0);
+}
+
+static int
+t_probe(const struct test *t, struct quark_queue_attr *qa)
+{
+	struct quark_queue	qq;
+
+	if (quark_queue_open(&qq, qa) != 0)
+		err(1, "%s: quark_queue_open", t->name);
+	quark_queue_close(&qq);
+
+	return (0);
+}
+
+static int
+t_fork_exec_exit(const struct test *t, struct quark_queue_attr *qa)
+{
+	struct quark_queue		 qq;
+	struct quark_event		 qev;
+	const struct quark_process	*qp;
+	pid_t				 child;
+	struct args			*args;
+	char				 cwd[PATH_MAX];
+
+	if (quark_queue_open(&qq, qa) != 0)
+		err(1, "%s: quark_queue_open", t->name);
+
+	child = fork_exec_nop();
+	if (drain_for_pid(&qq, child, &qev) != 0)
+		err(1, "drain_for_pid");
+
+	/* check qev.events */
+	assert(qev.events & QUARK_EV_FORK);
+	assert(qev.events & QUARK_EV_EXEC);
+	assert(qev.events & QUARK_EV_EXIT);
+	/* check qev.process */
+	qp = qev.process;
+	assert(qp != NULL);
+	assert(qp->flags & QUARK_F_EXIT);
+	assert(qp->flags & QUARK_F_COMM);
+	assert(qp->flags & QUARK_F_FILENAME);
+	assert(qp->flags & QUARK_F_CMDLINE);
+	assert(qp->flags & QUARK_F_CWD);
+	assert((pid_t)qp->pid == child);
+	assert((pid_t)qp->proc_ppid == getpid());
+	assert(qp->proc_time_boot > 0); /* XXX: improve */
+	assert(qp->proc_uid == getuid());
+	assert(qp->proc_gid == getgid());
+	assert(qp->proc_suid == geteuid());
+	assert(qp->proc_sgid == getegid());
+	assert(qp->proc_euid == geteuid());
+	assert(qp->proc_egid == getegid());
+	assert((pid_t)qp->proc_pgid == getpgid(0));
+	assert((pid_t)qp->proc_sid == getsid(0));
+	/* check capabilities */
+	/* XXX assumes we're root */
+	assert(qp->proc_cap_inheritable == 0);
+	/*
+	 * We don't know the exact set since it varies from kernel,
+	 * improve this in the future.
+	 */
+	assert(qp->proc_cap_effective != 0);
+	assert(qp->proc_cap_permitted != 0);
+	/* check entry leader */
+	/*
+	 * XXX TODO This depends how we're running the test, if we're over ssh
+	 * it will show ssh, if not it will show init and whatnot, for assert
+	 * that it is not unknown at least.
+	 */
+	assert(qp->proc_entry_leader != 0);
+	assert(qp->proc_entry_leader_type != QUARK_ELT_UNKNOWN);
+	/* XXX TODO check tty_major and tty_minor for self in the future */
+#if 0
+	assert(qp->proc_tty_major != QUARK_TTY_UNKNOWN);
+	assert(qp->proc_tty_minor != 0);
+#endif
+	/* check strings */
+	assert(!strcmp(qp->comm, "exe"));
+	assert(!strcmp(qp->filename, "/proc/self/exe"));
+	/* check args */
+	args = args_make(qp);
```

Comment:
where does `args_make` come from?

Reply:
that's a bit embarrassing, but it's from `quark.h`, I've been meaning to rename it:

```
/* XXX terrible name XXX */
struct args {
	char		*buf;
	size_t		 buf_len;
	int		 argc;
	const char	*argv[];
};
```